src/ui.h

```diff
@@ -3,19 +3,58 @@
 
 #include "enums.h"
 #include "output.h"
-#include <stdlib.h>
 #include "cursesdef.h"
+
+#include <stdlib.h>
+#include <unordered_map>
+#include <memory>
 ////////////////////////////////////////////////////////////////////////////////////
-/*
- * uimenu constants
- */
-const int UIMENU_INVALID = -1024;
-const int MENU_ALIGN_LEFT = -1;
-const int MENU_ALIGN_CENTER = 0;
-const int MENU_ALIGN_RIGHT = 1;
-const int MENU_WIDTH_ENTRIES = -2;
-const int MENU_AUTOASSIGN = -1;
+class ui_element;
 
+enum ui_element_type {
+    UI_ELEMENT_TEXT = 0,
+    UI_ELEMENT_INPUT,
+    UI_ELEMENT_PROMPT,
+    NUM_UI_ELEMENTS
+};
+
+// contains information about the newly created element, such as id and type
+struct ui_element_info {
+    ui_element_type type;   // type of element
+    std::string name;       // name of the element (used to refer to the element)
+    point size;             // size of the element
+    point pos;              // position of element
+    WINDOW *win;            // curses window handle
```

Comment:
If you want to be lazy and/or clever, you should use a `WINDOW_PTR` (see output.h) instead of a raw pointer. It makes sure the window gets deleted when the pointer goes out of scope. Similar for other instances of `WINWDW*`. However, you have to be careful when the pointer is shared, `WINDOW_PTR` is a `unique_ptr` and can not be shared.